src/mame/machine/archimedes_keyb.cpp

```diff
@@ -0,0 +1,337 @@
+}
+
+void archimedes_keyboard_device::device_timer(emu_timer &timer, device_timer_id id, int param, void *ptr)
+{
+	scan_keyb();
+}
+
+void archimedes_keyboard_device::tra_complete()
+{
+}
+
+void archimedes_keyboard_device::rcv_complete()
+{
+	receive_register_extract();
+	exec_command(get_received_char());
+}
+
+void archimedes_keyboard_device::tra_callback()
+{
+	m_kout(transmit_register_get_data_bit());
+}
+
+void archimedes_keyboard_device::exec_command(uint8_t command)
+{
+	LOG("%s: keyboard command %02x\n", machine().describe_context(), command);
+
+	if (command == 0xff)         // HRST: keyboard reset
+	{
+		transmit_register_setup(0xff);
+		m_keyb_enable  = false;
+		m_mouse_enable = false;
+		m_data = 0;
+	}
+	else if (command == 0xfe)    // RAK1: response in reset protocol
+		transmit_register_setup(0xfe);
+	else if (command == 0xfd)    // RAK2: response in reset protocol
+		transmit_register_setup(0xfd);
+	else if (command == 0x3f)    // BACK: ack for first data byte
+		transmit_register_setup((m_data >> 8) & 0xff);
+	else if (command == 0x20)    // RQID: request for keyboard ID
+		transmit_register_setup(0x81);
+	else if (command == 0x21)    // PRST: NOP
+	{
+	}
+	else if ((command & 0xf8) == 0x00)    // LEDS: set LEDs on/off
+	{
+		// -x-- Scroll lock
+		// --x- Num lock
+		// ---x Caps lock
+		m_leds[0] = BIT(command, 0);
+		m_leds[1] = BIT(command, 1);
+		m_leds[2] = BIT(command, 2);
+	}
+	else if ((command & 0xfc) == 0x30)    // NACK/SACK/MACK/SMAK: last data byte ack
+	{
+		m_data = 0;
+		m_keyb_enable = BIT(command, 0);
+		m_mouse_enable = BIT(command, 1);
+		scan_keyb();
+	}
+	else
+		logerror("%s: unknown keyboard command %02x\n", machine().describe_context(), command);
+}
+
+
+INPUT_CHANGED_MEMBER(archimedes_keyboard_device::update_mouse_input)
+{
+	m_mouse_x = (int16_t)m_mouse[0]->read();
+	m_mouse_y = (int16_t)m_mouse[1]->read();
+}
+
+void archimedes_keyboard_device::scan_keyb()
+{
+	if (!is_transmit_register_empty())
+		return;
+
+	if (m_keyb_enable)
+	{
+		for (int i = 0; i < 8; i++)
+		{
+			uint16_t data = m_keyboard[i]->read();
+			for (int b = 0; b < 16; b++)
+			{
+				int key_id = i * 16 + b;
+				bool new_state = BIT(data, b);
+				if (new_state != m_states[key_id])
+				{
+					if (new_state)
+						m_data = ((b | 0xc0) << 8) | (i | 0xc0);
+					else
+						m_data = ((b | 0xd0) << 8) | (i | 0xd0);
+
+					m_states[key_id] = new_state;
+					transmit_register_setup(m_data & 0xff);
+					return;
+				}
+			}
+		}
+	}
+
+	// sends mouse data only if there is no keyboard data
+	if (m_mouse_enable && (m_mouse_x != 0 || m_mouse_y != 0))
+	{
+		// keep in range -64 / +63
+		if (m_mouse_x < -64) m_mouse_x = -64;
+		if (m_mouse_y < -64) m_mouse_y = -64;
+		if (m_mouse_x >  63) m_mouse_x =  63;
+		if (m_mouse_y >  63) m_mouse_y =  63;
+
+		m_data = ((m_mouse_y & 0x7f) << 8) | (m_mouse_x & 0x7f);
+		transmit_register_setup(m_data & 0xff);
+		m_mouse_x = m_mouse_y = 0;
```

Comment:
Does it really saturate like this?  I ask because in most cases, the internal counter in the mouse has more range than the data frame.  This allows for a brief, rapid movement to track correctly.  For example Mouse Systems protocol devices typically have 10-bit or 12-bit counters internally but only 7-bit report values.  Sources like https://github.com/mamedev/mame/blob/mame0227/src/devices/bus/rs232/hlemouse.cpp deal with reporting the mouse movement in multiple frames.

Reply:
The keyboard MCU decodes the mouse quadrature signal and maintains a signed 7-bit counter, stopping to count in case of overflow / underflow (page 12 of the Archimedes 500 Service Manual).